### ISSUE TITLE
Goal: Ports to more architectures

### ISSUE THREAD

**Issue by lukego** [raptorjit/raptorjit#137]
RaptorJIT currently supports only x86-64. This is the dominant platform in server networking. However, the world moves forwards and in the future we will want to support more architectures like ARM64. Here are my thoughts on how to go about that.

First, the reason that I removed the other platforms is maintainability. If the VM is hand-written in assembler for a half-dozen architecture then it is very difficult to modify. Likewise when it is #ifdef'd to work with either a 32-bit or a 64-bit address space. This is especially true when there is no CI infrastructure providing testing and benchmark coverage for the platforms. The initial broad platform support inherited from LuaJIT makes changing the VM very expensive, and this expense is not justified in the present server world with ~99% market share held by x86-64 CPUs.

The next step that I want to take is to rewrite the x86-64 bytecode interpreter from assembler into C. This will make it more feasible to implement potential killer features like a 96-bit word size (#93). I have done some preliminary hacking in this direction and it is enough to convince me that a rewrite in C is the way forward and that the tricky issues like unwinding a mixture of Lua/C stack frames are actually straightforward in C when following the PUC Lua approach. (Thanks @igelhaus for all the help on working this part out over at #110!)

Once the VM is written in C then the incremental cost of new platform support will be much less. The JIT backend will be needed but we won't need thousands of lines of hand-coded VM. This means that new ports will not be a drag on evolving the VM. I do propose that we stay somewhat conservative nevertheless, for example restricting ourselves to 64-bit architectures only.

There will be some overhead to running a C bytecode interpreter rather than an assembler one. I believe this will be massively outweighed by the benefits. One reason is that optimized programs don't spend any significant amount of time running the interpreter at all ("don't optimize the idle loop.") One more is that there are massive potential performance gains to be had from VM evolution, e.g. by supporting unboxed pointers and 64-bit integers so that `CNEWI` instructions don't heap-allocate and don't critically have to be sunk. (This last point is one of the worst problems in optimizing code today, I am *so* read to move on from giving a fig about whether my CNEWIs have always sunk or not... but I digress.)

TLDR; RaptorJIT can support all important 64-bit server platforms in the future.

**Comment by raj2569** [comment]
On Fri, Dec 1, 2017 at 3:23 PM, Luke Gorrie <notifications@github.com>
wrote:

Do you ever see raptorjit moving to using an 128 bit word? I am not sure
how feasible or even necessary such a word may be, but with ipv6 becoming
more popular day by day, a networking app like Snabb will have to deal with
128bit words quite frequently. So an efficient mechanism to handle ipv6
address would be nice.

**Comment by lukego** [comment]
@raj2569 You read my mind! :-)

I have actually been thinking about the implications of having a word size that accommodates all the major datatypes in networking, the largest common one of which is the 128-bit IPv6 address as you say.

This is technically distinct from the 96-bit word size. That's because the 96-bit is a 64-bit value with a 32-bit type. This means that the value part can still fit into a 64-bit register and only needs to be expanded to 96 bits when stored in memory. Have to think about whether this approach could be generalized to larger values too (e.g. using 128/256/512-bit xmm/ymm/zmm registers.)

Have to think more about this! Could also be that we can solve this one at the Lua library level ("design pattern") but I am not sure about that.

**Comment by szbnwer** [comment]
something to keep an eye on:
https://www.wdc.com/about-wd/newsroom/press-room/2017-11-28-western-digital-to-accelerate-the-future-of-next-generation-computing-architectures-for-big-data-and-fast-data-environments.html

cant the asm and c variants coexists for experimenting? this makes moving forward harder, but good to not leave the coast too far before know the real results. i think you will do modifications here and there, and the asm from lj wont fit to go back easily in case of need. i think the worse thing around software development in general is the hurrying forward instead of make everything slowly the best (cuz the way of business going). hard parts can be always well documented instead of making it easier and worse. most important is to document the "dont touch, magic" parts for the reasons why that particular solution is the best, how it works, what else have been tried that werent good for much if anything, and any additional resources that are enlightening. the problem with hardcore codes is not that they are hardcore, but that the info that made them become themselves disappeared on the road, and mostly known by max 1 person

**Comment by pygy** [comment]
@lukego do you intend to use C extensions to make a direct threaded interpreter? If not, aren't you concerned about potential loss of performance in a post-Spectre world? My understanding Spectre is hazy, but IIUC the indirect branch predictors that's present in the latest Intel chips may have to be turned off, giving once again a perf advantage to direct threaded interpreters.

**Comment by lukego** [comment]
@szbnwer I usually use Linus's distributed development model for big changes. Rewriting the VM in C would be done on a branch. This branch could be used to experiment with new optimizations. If these optimizations work well then they would be merged upstream (and bring the C VM with them.)

So the master branch always represents the current "proven" approach and everything else exists on a publicly accessible branch where people can experiment with it.

**Comment by lukego** [comment]
@pygy I would be very interested in any citations in this area. I am hoping that the performance hit will ultimately only be paid when switching between multiple privilege levels on the same core (e.g. system calls) and that the JIT will be unaffected. We will see though. I am waiting for the dust to settle now.

**Comment by pygy** [comment]
@lukego My comment is based on http://article.gmane.org/gmane.comp.lang.lua.general/75426 (a post where Mike Pall details why he wrote the interpreter in ASM), and the fact that there are forks of the PUC Lua interpreter that use (IIRC) computed gotos to improve speed, but the speed benefits disappeared for the latest Intel chips thanks to improvements to their branch predictor.

I'm not sure those improvements are what lead to Spectre, but they seem quite similar.

**Comment by darius** [comment]
Re: Spectre, I wonder about the tradeoffs in trying to ward it off. Of the integrity/confidentiality/availability triad, it affects confidentiality, but that was already quite leaky in a tracing JIT compiler, right? Might it make sense to just warn that this compiler does not try to plug side channels? Or supply a mode that does try, by staying in the interpreter, etc.?

**Comment by lukego** [comment]
My take is that interpreter performance is almost completely irrelevant. It's like swap memory on a Linux machine. Sure, it's nice that it exists as a fallback, but you would never use it in a performance sensitive application. Instead you would fix all "NYIs" to keep the code in JIT mode and you would buy enough RAM to never spill your working set onto swap. You are wasting your time if you try to maximize performance by working on the interpreter or the virtual memory swapping subsystem because they are the wrong tools for the job.

Mike successfully built a mystique for LuaJIT based on that email - "holy cow it's written by a genius who can kick the C compiler's ass!" - but as an application developer it only causes me grief.

(FreeBSD kicks Linux's ass at virtual memory swapping performance, and always has, but nobody is going to adopt it for that reason.)

I am still not aware of any specific changes to the virtual machine that should be motivated by Spectre. I would like a citation. Certainly I would say that tracing JIT performance depends absolutely on speculative execution - guards don't block real work - and if that would disappear then it's time to start programming in C again. But I'm assuming those CPU features are here to stay until proven otherwise.

**Comment by lukego** [comment]
(Sorry to rant @pygy :-) I am just really looking forward to putting the complex interpreter code behind me and optimizing it for maintainability instead.)

**Comment by lukego** [comment]
I must confess that I am not tracking the Spectre and Meltdown issues in detail and so I could be missing something important. Just now my mental model is that these issues are about leaking information between multiple privilege domains running on the same CPU core e.g. reading memory outside a sandbox. That would not seem directly relevant to RaptorJIT since the code is all running at the same privilege level i.e. we are not implementing any sandboxes.

I suppose that it would be an issue if you would use a "sanitized" Lua environment to run untrusted code in the RaptorJIT VM. This code could use side-channel attacks to read outside the sandbox. This might be a legitimate concern. However, I am suspicious of whether this is a realistic approach to security from the beginning and I wouldn't recommended doing this.

**Comment by pygy** [comment]
No worries @lukego I can totally see where you're coming from. If interpreter perf is not a priority then you can totally ignore my comment :-)

My understanding of Spectre and Meltdown is pretty superficial too.

AFAIU, Meltdown is the fact the memory protection is not enforced during speculation, which enables a rogue process to read RAM it isn't allowed to then, depending on the protected content, load some data it has access to in the cache. It avoids page faults by tricking the branch predictor into believing that some actually false condition is likely to be true. After that, it measures the time required to read the data that was put in cache.

Spectre is about tricking the branch predictor into jumping into arbitrary code and use return-oriented programming while speculating then leave info about what happened by loading stuff in the cache. It is problematic for sandboxes like browsers, but also for processes that can be manipulated from the outside like the OS kernel. There is a Spectre PoC that uses the Linux eBPF interpreter to execute arbitrary kernel code (once again, defeating memory protection). There are also JS PoC that extract info from other tabs.

IIUC, the Retpoline compiler fix prevents processors from doing ROP while speculating by preventing indirect branch prediction when returning from a function. It makes a program Spectre-resistant.

Adding Retpoline to the JIT may or may not be a good idea, I don't know enough to judge.

**Comment by lukego** [comment]
Thanks for sharing @pygy. Sorry to fly off the handle whenever a discussion leads towards spending more time thinking about the interpreter code :).

In the Snabb context if we want to write raw physical memory we just write something like `S.mmap('/dev/mem')[0x424242] = 0` because we are *supposed* to have complete control of the whole machine anyway :).

**Comment by darius** [comment]
A program could tell what tabs you had open even before Spectre came along (https://github.com/defuse/flush-reload-attacks). Spectre just opens the confidentiality breach wider.

**Comment by szbnwer** [comment]
@lukego 

1: ur method is totally fine, but ive meant in my prev msg that you are mixing codes from replacing asm with c and from other targeted features, and in case of need it could be a harder nut to go back to asm when u already have other modifications upon these changes, so probably its not an option to just copy-paste back the asm, however probably not even that much hard til you can follow up ur modifications - and take this only as something to consider... i think ill try and like rj anyhow in the sooner or later future (im always in lack of time and i think ive already shrunk :D :/ )

2: isnt one of the aims of using rj is to use it in the hosting area? if only for backend stuffs, then meltdown/spectre/rowhammer/anything else cant just harm anything, but if its related anyhow to host providing, then these are creepy for me, however i dunno where this came, just i have a deep impression that ive seen something similar previously, and i have no enough deep understanding above these vulnerabilities, so i can be more creeped than its needed :D 

@pygy 
may i ask u to enlighten my confused mind around the potential of these vulners like what they actually could have done (for linux and intel) via updates for firmwares/kernel/browsers against these, and whats the remaining attack vector for them for an unprivileged process for a browser and for a shared hosting provider like how far they can see/manipulate other things than intended to? and so many thanks im trying to figure these out but the mess in my head just grows, and u seems to be like some1 who maybe know the straight answer for these or some of these and therefore can explain better than random techy articles... (i only dare to ask it here cuz i think it can beneficial for the others around as well :D ) ive got some basic messy understanding about these and things around, so as short answers as 'meltdown can reach this far but not that if you have the updates' and the same for spectre, and the most important for me is the potential in browsers, cuz thats maybe the only source i can get anything wrong (i have 1147 opened tabs currently (even some undone lessons about meltdown and spectre) X'D i really cant follow up myself....), however i appreciate anything more and thx once again!
